文章封面图片链接包含空格时，`${thumbnail.gen(cover, 's')}` 报错

### 系统信息

- 外部访问地址: https://2a1e-14-155-63-83.ngrok-free.app
- 启动时间: 2024-09-23 13:42
- 版本: 2.19.3
- 构建时间: 2024-09-10 17:46
- Git Commit: b6ec7b9
- Java: IBM Semeru Runtime Open Edition / 21.0.4+7-LTS
- 数据库: PostgreSQL / 15.4 (Debian 15.4-2.pgdg120+1)
- 操作系统: Linux / 5.15.153.1-microsoft-standard-WSL2
- 已激活主题: [Theme Joe3 1.4.0](https://github.com/jiewenhuang/halo-theme-joe3.0)
- 已启动插件:
  - [豆瓣 1.2.1](https://github.com/chengzhongxue/plugin-douban)
  - [对象存储（Amazon S3 协议） 1.11.0](https://github.com/halo-dev/plugin-s3)
  - [代码注入 1.0.0-SNAPSHOT](https://github.com/halo-sigs/plugin-injection)
  - [MaxKB 小助手 1.1.3](https://www.halo.run/store/apps/app-aWHcE)
  - [代码注入增强插件 1.0.0-SNAPSHOT](https://github.com/halo-sigs/plugin-codeInjection)
  - [编辑器超链接卡片 1.0.3](https://github.com/halo-sigs/plugin-editor-hyperlink-card)
  - [Data Studio（数据工厂） 1.0.0-alpha.2](https://github.com/halo-sigs/plugin-data-studio)
  - [插件快速开始模板 1.0.0-SNAPSHOT](https://github.com/halo-dev/plugin-hellopage)
  - [静态网页服务 1.0.0-alpha.1](https://github.com/ruibaby/plugin-static-pages)
  - [Docsme 1.0.0-alpha.7](https://www.halo.run/store/apps/app-yffxw)
  - [SEO 工具集 1.0.6](https://www.halo.run/store/apps/app-FNGbT)
  - 联系表单 1.0.2
  - Reflect Halo 微信小程序配套插件 1.0.0-SNAPSHOT
  - [应用市场 1.5.0](https://www.halo.run/store/apps/app-VYJbF)
  - [社交 IAM 认证 1.0.2-SNAPSHOT](https://www.halo.run/store/apps/app-IXZkJ)
  - [AI 助手 1.0.0](https://www.halo.run/store/apps/app-riNgb)
  - [文章限制阅读 Pro 1.4.0-beta.2](https://www.halo.run/store/apps/app-TzRqd)
  - [Markdown / HTML 内容块 1.0.1](https://github.com/halo-dev/plugin-hybrid-edit-block)
  - [WangEditor 1.0.0](https://gitee.com/handsome-npc/halo-wangEditor-plugin)
  - [微信公众号发布 1.1.2-SNAPSHOT](https://github.com/f2ccloud/plugin-platforms-sync)
  - [站点迁移 1.2.0-SNAPSHOT](https://github.com/halo-sigs/plugin-migrate)
  - [StackEdit 1.1.1](https://github.com/halo-sigs/plugin-stackedit)
  - [ByteMD 1.4.0](https://github.com/ruibaby/plugin-bytemd)
  - [图库管理 1.3.0](https://github.com/halo-sigs/plugin-photos)
  - [链接管理 1.3.3](https://github.com/halo-sigs/plugin-links)
  - [LDAP 认证 1.0.0-SNAPSHOT](https://github.com/f2ccloud/plugin-ldap-login)
  - [Sitemap 1.1.2](https://github.com/halo-dev/plugin-sitemap)
  - [Unsplash 1.0.0](https://github.com/halo-sigs/plugin-unsplash)
  - [reveal.js 幻灯片编辑器 1.0.0-rc.1](https://github.com/wan92hen/plugin-revealjs)
  - [瞬间 1.5.1](https://github.com/halo-sigs/plugin-moments)
  - [OAuth2 认证 1.1.0](https://github.com/halo-sigs/plugin-oauth2)
  - [RSS 1.1.1](https://github.com/halo-sigs/plugin-feed)
  - [评论组件 2.4.0](https://github.com/halo-dev/plugin-comment-widget)
  - [搜索组件 1.5.0](https://github.com/halo-dev/plugin-search-widget)
  - [highlight.js 代码高亮 1.2.2](https://github.com/halo-sigs/plugin-highlightjs)

### 使用的哪种方式运行？

Docker

### 发生了什么？

文章封面使用包含空格的图片链接时，主题模板通过 `${thumbnail.gen(cover, 's')}` 方法配置响应式图片时报错。


### 复现步骤

1. 安装并使用适配了文章封面响应式图片的主题，例如 Theme Joe3 1.4.0
1. 准备一个名字包含空格的图片文件，例如 `PixPin (1).png`
2. 上传该图片到附件列表
3. 设置文章封面，从附件库选择该图片时，最终 url 进行了编码处理 `/upload/test/PixPin%20(1).png`，此时访问主题页面正常
4. 设置文章封面时手动输入图片链接 `/upload/test/PixPin (1).png`，此时访问主题页面报错

### 相关日志输出

```shell
Caused by: org.thymeleaf.exceptions.TemplateProcessingException: Exception evaluating SpringEL expression: "thumbnail.gen(cover, 's')" (template: "modules/macro/post_item" - line 26, col 15)
        at org.thymeleaf.spring6.expression.SPELVariableExpressionEvaluator.evaluate(SPELVariableExpressionEvaluator.java:292) ~[thymeleaf-spring6-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at run.halo.app.theme.ReactiveSpelVariableExpressionEvaluator.evaluate(ReactiveSpelVariableExpressionEvaluator.java:29) ~[classes/:2.19.3]
        at org.thymeleaf.standard.expression.VariableExpression.executeVariableExpression(VariableExpression.java:166) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.SimpleExpression.executeSimple(SimpleExpression.java:66) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:109) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.AdditionExpression.executeAddition(AdditionExpression.java:89) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.ComplexExpression.executeComplex(ComplexExpression.java:62) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:112) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.expression.Expression.execute(Expression.java:138) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.processor.StandardDefaultAttributesTagProcessor.processDefaultAttribute(StandardDefaultAttributesTagProcessor.java:165) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.processor.StandardDefaultAttributesTagProcessor.process(StandardDefaultAttributesTagProcessor.java:98) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.util.ProcessorConfigurationUtils$ElementTagProcessorWrapper.process(ProcessorConfigurationUtils.java:633) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.ProcessorTemplateHandler.handleStandaloneElement(ProcessorTemplateHandler.java:918) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.StandaloneElementTag.beHandled(StandaloneElementTag.java:228) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.Model.process(Model.java:282) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.ProcessorTemplateHandler.handleStandaloneElement(ProcessorTemplateHandler.java:1204) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.StandaloneElementTag.beHandled(StandaloneElementTag.java:228) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.Model.process(Model.java:282) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.Model.process(Model.java:290) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.IteratedGatheringModelProcessable.processIterationModel(IteratedGatheringModelProcessable.java:368) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.IteratedGatheringModelProcessable.process(IteratedGatheringModelProcessable.java:294) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.ProcessorTemplateHandler.handleCloseElement(ProcessorTemplateHandler.java:1640) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.CloseElementTag.beHandled(CloseElementTag.java:139) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.Model.process(Model.java:282) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.ProcessorTemplateHandler.handleOpenElement(ProcessorTemplateHandler.java:1587) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.OpenElementTag.beHandled(OpenElementTag.java:205) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.Model.process(Model.java:282) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.ProcessorTemplateHandler.handleOpenElement(ProcessorTemplateHandler.java:1587) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.engine.TemplateHandlerAdapterMarkupHandler.handleOpenElementEnd(TemplateHandlerAdapterMarkupHandler.java:304) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.templateparser.markup.InlinedOutputExpressionMarkupHandler$InlineMarkupAdapterPreProcessorHandler.handleOpenElementEnd(InlinedOutputExpressionMarkupHandler.java:278) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.standard.inline.OutputExpressionInlinePreProcessorHandler.handleOpenElementEnd(OutputExpressionInlinePreProcessorHandler.java:186) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.thymeleaf.templateparser.markup.InlinedOutputExpressionMarkupHandler.handleOpenElementEnd(InlinedOutputExpressionMarkupHandler.java:124) ~[thymeleaf-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        at org.attoparser.HtmlElement.handleOpenElementEnd(HtmlElement.java:109) ~[attoparser-2.0.7.RELEASE.jar:2.0.7.RELEASE]
        at org.attoparser.HtmlMarkupHandler.handleOpenElementEnd(HtmlMarkupHandler.java:297) ~[attoparser-2.0.7.RELEASE.jar:2.0.7.RELEASE]
        at org.attoparser.MarkupEventProcessorHandler.handleOpenElementEnd(MarkupEventProcessorHandler.java:402) ~[attoparser-2.0.7.RELEASE.jar:2.0.7.RELEASE]
        at org.attoparser.ParsingElementMarkupUtil.parseOpenElement(ParsingElementMarkupUtil.java:159) ~[attoparser-2.0.7.RELEASE.jar:2.0.7.RELEASE]
        at org.attoparser.MarkupParser.parseBuffer(MarkupParser.java:710) ~[attoparser-2.0.7.RELEASE.jar:2.0.7.RELEASE]
        at org.attoparser.MarkupParser.parseDocument(MarkupParser.java:301) ~[attoparser-2.0.7.RELEASE.jar:2.0.7.RELEASE]
        ... 17 common frames omitted
Caused by: java.lang.IllegalArgumentException: Illegal character in path at index 19: /upload/test/PixPin (1).png
        at java.base/java.net.URI.create(Unknown Source) ~[na:na]
        at run.halo.app.theme.finders.impl.ThumbnailFinderImpl.gen(ThumbnailFinderImpl.java:18) ~[classes/:2.19.3]
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source) ~[na:na]
        at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
        at org.springframework.expression.spel.support.ReflectiveMethodExecutor.execute(ReflectiveMethodExecutor.java:142) ~[spring-expression-6.1.12.jar:6.1.12]
        at org.springframework.expression.spel.ast.MethodReference.getValueInternal(MethodReference.java:125) ~[spring-expression-6.1.12.jar:6.1.12]
        at org.springframework.expression.spel.ast.MethodReference$MethodValueRef.getValue(MethodReference.java:401) ~[spring-expression-6.1.12.jar:6.1.12]
        at org.springframework.expression.spel.ast.CompoundExpression.getValueInternal(CompoundExpression.java:97) ~[spring-expression-6.1.12.jar:6.1.12]
        at org.springframework.expression.spel.ast.SpelNodeImpl.getValue(SpelNodeImpl.java:114) ~[spring-expression-6.1.12.jar:6.1.12]
        at org.springframework.expression.spel.standard.SpelExpression.getValue(SpelExpression.java:338) ~[spring-expression-6.1.12.jar:6.1.12]
        at org.thymeleaf.spring6.expression.SPELVariableExpressionEvaluator.evaluate(SPELVariableExpressionEvaluator.java:265) ~[thymeleaf-spring6-3.1.2.RELEASE.jar:3.1.2.RELEASE]
        ... 72 common frames omitted
Caused by: java.net.URISyntaxException: Illegal character in path at index 19: /upload/test/PixPin (1).png
        at java.base/java.net.URI$Parser.fail(Unknown Source) ~[na:na]
        at java.base/java.net.URI$Parser.checkChars(Unknown Source) ~[na:na]
        at java.base/java.net.URI$Parser.parseHierarchical(Unknown Source) ~[na:na]
        at java.base/java.net.URI$Parser.parse(Unknown Source) ~[na:na]
        at java.base/java.net.URI.<init>(Unknown Source) ~[na:na]
        ... 83 common frames omitted
```


### 附加信息

_No response_

--- a/application/src/main/java/run/halo/app/theme/finders/impl/ThumbnailFinderImpl.java
+++ b/application/src/main/java/run/halo/app/theme/finders/impl/ThumbnailFinderImpl.java
@@ -2,12 +2,14 @@ package run.halo.app.theme.finders.impl;
 
 import java.net.URI;
 import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
 import reactor.core.publisher.Mono;
 import run.halo.app.core.attachment.ThumbnailService;
 import run.halo.app.core.attachment.ThumbnailSize;
 import run.halo.app.theme.finders.Finder;
 import run.halo.app.theme.finders.ThumbnailFinder;
 
+@Slf4j
 @Finder("thumbnail")
 @RequiredArgsConstructor
 public class ThumbnailFinderImpl implements ThumbnailFinder {
@@ -15,8 +17,14 @@ public class ThumbnailFinderImpl implements ThumbnailFinder {
 
     @Override
     public Mono<String> gen(String uriStr, String size) {
-        return thumbnailService.generate(URI.create(uriStr), ThumbnailSize.fromName(size))
+        return Mono.fromSupplier(() -> URI.create(uriStr))
+            .flatMap(uri -> thumbnailService.generate(uri, ThumbnailSize.fromName(size)))
             .map(URI::toString)
+            .onErrorResume(Throwable.class, e -> {
+                log.debug("Failed to generate thumbnail for [{}], error: [{}]", uriStr,
+                    e.getMessage());
+                return Mono.just(uriStr);
+            })
             .defaultIfEmpty(uriStr);
     }
 }

--- a/application/src/test/java/run/halo/app/theme/finders/impl/ThumbnailFinderImplTest.java
+++ b/application/src/test/java/run/halo/app/theme/finders/impl/ThumbnailFinderImplTest.java
@@ -1,0 +1,54 @@
+package run.halo.app.theme.finders.impl;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.net.URI;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import reactor.core.publisher.Mono;
+import reactor.test.StepVerifier;
+import run.halo.app.core.attachment.ThumbnailService;
+
+/**
+ * Tests for {@link ThumbnailFinderImpl}.
+ *
+ * @author guqing
+ * @since 2.20.0
+ */
+@ExtendWith(MockitoExtension.class)
+class ThumbnailFinderImplTest {
+
+    @Mock
+    ThumbnailService thumbnailService;
+
+    @InjectMocks
+    ThumbnailFinderImpl thumbnailFinder;
+
+    @Test
+    void shouldNotGenWhenUriIsInvalid() {
+        thumbnailFinder.gen("invalid uri", "l")
+            .as(StepVerifier::create)
+            .expectNext("invalid uri")
+            .verifyComplete();
+
+        verify(thumbnailService, times(0)).generate(any(), any());
+    }
+
+    @Test
+    void shouldGenWhenUriIsValid() {
+        when(thumbnailService.generate(any(), any()))
+            .thenReturn(Mono.just(URI.create("/test-thumb.jpg")));
+        thumbnailFinder.gen("/test.jpg", "l")
+            .as(StepVerifier::create)
+            .expectNext("/test-thumb.jpg")
+            .verifyComplete();
+
+        verify(thumbnailService).generate(any(), any());
+    }
+}

